syntax = "proto3";
package csce438;
import "google/protobuf/timestamp.proto";

// ------------------------------------------------------------
// The coordinator service definition
// These are the different functionalities exposed by service
// ------------------------------------------------------------

// NOTE: Most of what is in this file will not be needed because communication between follower synchronizer processes uses RabbitMQ instead of gRPC.

service CoordService {
    rpc Heartbeat (ServerInfo) returns (Confirmation) {}
    rpc GetServer (ID) returns (ServerInfo) {}
    rpc GetSlave (ID) returns (ServerInfo) {}
    rpc GetAllFollowerServers (ID) returns (ServerList) {}
    rpc GetFollowerServer (ID) returns (ServerInfo) {}
}

// server info message definition
message ServerInfo {
    int32 serverID = 1;
    string hostname = 2;
    string port = 3;
    string type = 4;
    int32 clusterID = 5;
    bool isMaster = 6;
}

// ServerList definition
message ServerList {
    repeated int32 serverID = 1;
    repeated string hostname = 2;
    repeated string port = 3;
    repeated string type = 4;
}

// confirmation message definition
message Confirmation {
    bool status = 1;
}

// id message definition
message ID {
    int32 id = 1;
}
